Add plugin ui view

ibedard16 opened the pull request:
This adds a component that gets shown whenever the "plugins" tab is pressed. Here's what it looks like currently:

![image](https://user-images.githubusercontent.com/13560916/79177989-0513a880-7dd2-11ea-9b13-0477cfc5bce5.png)

Apologies for my messy commit history. Was figuring things out.

(Note: this pulls in changes from https://github.com/MultimediaExtensibleGit/Runtime/pull/9, so that should be merged in first)

Commit:
Attempt to add plugins panel to panel list

Can't get it to display. Making commit for advice.

Commit:
Add horizontal & vertical layouts

Commit:
Actually fix layout

Kudos to Kyle

Commit:
Add handlers and refresh_panel()

Commit:
Merge remote-tracking branch 'origin/ui-handlers' into plugin-ui

Commit:
Fix mistakes in pluginspanel.py

Oops, had some mixed whitespace types.
Also, don't know what was going on with all of the "self" references. Should have been instance.

Also, javascritpt ternary syntax does not work in python

Commit:
Fix manager file

Remove code I had added to attempt to test the plugins panel.

Add an implementation for the open_manage_plugins function

Commit:
Remove manager property from panel initializer

Commit:
Fix plugins panel

Kudos to Jake for fixing my syntax errors... like trying to reference a class I never imported, or calling methods that didn't exist. (Seriously, why weren't those errors more explicit?)

Discovered the list of plugins needs to be cleared & re-populated each time the list is loaded, otherwise the list of plugins just grows each time its opened.

Commit:
Fix the logger

The logger's broke. This fixes it.

Commit:
Start add plugin panel implementation

Still not finalized yet, wanted to commit this for now for smaller commits (for review purposes).

Commit:
Make the add plugin panel fully functional

Lots of code, lots of code

Made it so that plugins can be installed in a variety of ways.

Commit:
Add functionality to enable/disable/uninstall plugin buttons

These were just placeholders before because I wasn't sure how to interface with the QTreeWidget. But now I know, and these aren't placeholders anymore.

Commit:
Keep unicode out of source file

I tried using `\u` syntax, but the encoding for the blue circle was `\u1F535`. For some reason, python didn't like handling a character encoding with 5 digits.

Either way, this keeps unicode out of our source files

kyletpugh commented on meg_runtime/app.py:
```diff
@@ -60,6 +62,26 @@ def get_panel(name):
         """Get a panel by name"""
         return None if not App.__instance else App.__instance.panel(name)
 
+    @staticmethod
+    def refresh_panel(name, **kwargs):
+        """Refresh a panel by name."""
+        instance = App.__instance
+        if instance is None:
+            return None
+        if name in App.PANELS:
+            try:
+                panel_ctor = getattr(ui, name)
+                panel_obj = panel_ctor(**kwargs)
+                instance._panels[panel_obj.get_name()] = panel_obj
+                return panel_obj
+            except Exception as e:
+                Logger.warning(f'MEG UI: {e}')
+                Logger.warning(f'MEG UI: Could not create panel "{name}"')
+        else:
+            Logger.warning(f'MEG UI: Panel does not exist')
+
+        return None
```
I think I am going to change the method of how the panels are populated some more with a few sets of handlers in the BasePanel. This seems inadequate and expensive to reread the widgets when they already exist. Say, I want to go back to the main panel through the back button, I don't expect that the panel will be reset but these changes do that. If I go to the main panel from say the menu or the start of the app, I do expect that the panel is reset. So, I was thinking on_start()/on_stop() to modify the panels when they are navigated to through changing the view (the other navigation will be stack based back and forward, etc) and on_show()/on_hide() to modify for when the panels visibility changes to adjust any changes that need updated only. Then we can have timers activated in on_show() and deactivated in on_hide() to periodically update the interface, like in the case of the plugins list/add panels. The on_start() and on_stop() can be used to clone the widgets that are loaded for the class, store them on the instance then remove them instead of entirely re-reading them, and you can make per instance changes and have multiple copies in the stack to go back and forth. Panels that do this would need to be uniquely named, since only one name can exist to store the panel. Just rambling thoughts about how the panels don't quite behave correctly, nothing really to be done, here.

kyletpugh commented on meg_runtime/ui/addpluginpanel.py:
```diff
@@ -0,0 +1,130 @@
+        messageBox = QtWidgets.QMessageBox()
+        messageBox.setIcon(QtWidgets.QMessageBox.Critical)
+        messageBox.setText(message)
+        messageBox.exec();
```
Need a new line at the end of the file.

kyletpugh commented on meg_runtime/ui/addpluginpanel.py:
```diff
@@ -0,0 +1,130 @@
+from meg_runtime.plugins import PluginManager
+from meg_runtime.ui.basepanel import BasePanel
+
+class AddPluginPanel(BasePanel):
```
Global definitions require two empty separator lines

kyletpugh commented on meg_runtime/app.py:
```diff
@@ -60,6 +62,26 @@ def get_panel(name):
+
     def name(self):
         """Get application name"""
         return App.APP_NAME
```
Need version() and also did I name that APP_NAME in App, lol? I think should have just App.NAME and App.VERSION (probably can change the name of App.APP_ICON_PATH to App.ICON_PATH as well). Also should have static method get_version() like get_name(), that calls the instance method.

Commit:
Fix typos, remove semicolons

kyletpugh commented on meg_runtime/ui/addpluginpanel.py:
```diff
@@ -0,0 +1,130 @@
+class AddPluginPanel(BasePanel):
+    """Setup the plugin panel."""
+
+    selected_file = None
```
I very much doubt that you want this to be a class variable and not an instance variable. You should be setting this on self instance in \_\_init\_\_(), then using self.selected_file.

kyletpugh commented on meg_runtime/ui/addpluginpanel.py:
```diff
@@ -0,0 +1,130 @@
+        self.url_radio_button.clicked.connect(self.enable_url_selection)
+        self.url_field = instance.findChild(QtWidgets.QLineEdit, 'urlField')
+
+        # reset form values
+        self.available_radio_button.setChecked(True)
+        self.available_plugin_list.clear()
+        self.file_label.setText('')
+        self.url_field.setText('')
+
+        # Add all available plugins to the available plugin list
+        available_plugins = PluginManager.get_all_available()
+        for plugin in available_plugins:
+            self.available_plugin_list.addTopLevelItem(QtWidgets.QTreeWidgetItem([
+                plugin.name(),
+                plugin.version(),
+                plugin.author(),
+                plugin.description()
+            ]))
```
This would benefit from the other method, then you could just reset the form in on_start().

kyletpugh commented on meg_runtime/ui/manager.py:
```diff
@@ -47,6 +60,58 @@ def get_instance(**kwargs):
+        """Open the about menu."""
+        instance = UIManager.get_instance()
+        desc = ('<center>'
+                '<h1>Multimedia Extensible Git</h1>'
+                '<p>Version 0.1</p>'
+                '</center>')
+        QtWidgets.QMessageBox.about(instance, "About MEG", desc)
```
This should use App.get_name() and App.get_version() (which needs created apparently). Also, the title should be `f'About {App.get_name()}'`.

kyletpugh commented on meg_runtime/ui/manager.py:
```diff
@@ -57,12 +122,12 @@ def setup(**kwargs):
     def return_to_main():
         """Return to the main panel"""
-        UIManager.change_view(App.get_panel('MainPanel'))
+        UIManager.change_view(App.refresh_panel('MainPanel'))
```
Seems like going back shouldn't reset, this should maybe be open_main_panel()?

kyletpugh commented on meg_runtime/ui/manager.py:
```diff
@@ -78,7 +143,6 @@ def send_changes(repo):
     @staticmethod
     def change_view(panel):
         """Change the current panel being viewed. """
-        # Reload the panel before changing the view
```
Put a useful comment in a place where it obviously is wrong... lol.

kyletpugh commented on meg_runtime/ui/pluginspanel.py:
```diff
@@ -0,0 +1,100 @@
+        self.plugin_list = instance.findChild(QtWidgets.QTreeWidget, 'pluginList')
+        self.plugin_list.itemSelectionChanged.connect(self.changeButtonStates)
+
+        self.refreshPluginList()
+
+    def refreshPluginList(self):
+        self.plugin_list.clear()
+        plugins = PluginManager.get_all()
+        for plugin in plugins:
+            self.plugin_list.addTopLevelItem(QtWidgets.QTreeWidgetItem([
+                chr(128309) if plugin.enabled() else chr(9898),
+                plugin.name(),
+                plugin.version(),
+                plugin.author(),
+                plugin.description()
+            ]))
+        
+        # disable buttons
+        self.changeButtonStates()
```
Same as the other panel, would benefit from the other method of handling the panel interactions.

kyletpugh commented on meg_runtime/ui/pluginspanel.py:
```diff
@@ -0,0 +1,100 @@
+        
+        selectedPluginName = selectedPluginItem.text(1)
+        return PluginManager.get(selectedPluginName)
+
+    
+
+
+
```
Fix your spacing in this file. Only need one empty line at end of file.

kyletpugh commented on meg_runtime/ui/repopanel.py:
```diff
@@ -19,37 +19,35 @@ def __init__(self, repo_url=None, repo_path=None, repo=None,**kwargs):
-
     def get_title(self):
         """Get the title of this panel."""
         instance = self.get_widgets()
```
No need to get the instance of widgets here.

kyletpugh commented on meg_runtime/ui/repopanel.py:
```diff
@@ -19,37 +19,35 @@ def __init__(self, repo_url=None, repo_path=None, repo=None,**kwargs):
         self._repo_path = repo_path
         self._repo = repo
         super().__init__(**kwargs)
+        self._branch_name_label.text = self.title
```
Do this in on_load() (or on_start() or on_show() later).

kyletpugh commented on meg_runtime/ui/pluginspanel.py:
```diff
@@ -0,0 +1,100 @@
+from meg_runtime.ui.basepanel import BasePanel
+from meg_runtime.ui.manager import UIManager
+
+class PluginsPanel(BasePanel):
```
Need two empty lines before global definitions.

Commit:
Remove excess whitespace, refactor into methods for readability

Apparently there's a rule in python that prevents methods from having whitespace in them. I think that rule is dumb.

Since I had to make this change, I decided to refactor parts I had previously separated with whitespace into their own methods, for readability.
